File: hospitalManagement/Employee.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
namespace hospitalManagement
{
    internal abstract class Employee : IBill
    {
        //Field
        private string id;

        // Infor field
        private string firstName;

        private string lastName;

        private int age;

        private int gentle;

        private string description;

        private string address;

        private string email;

        private string phone;

        //  Management field
        private string facultyId;

        private string departmentId;

        private bool state;


        private ComeAndLeaveDate admissionDates;

        private Salary salaries;

        // Dynamic field
        // Properties
        public string Id { get => id; set => id = value; }
        public string FirstName { get => firstName; set => firstName = value; }
        public string LastName { get => lastName; set => lastName = value; }
        public int Age { get => age; set => age = value; }
        public int Gentle { get => gentle; set => gentle = value; }
        public string Description { get => description; set => description = value; }
        public string Address { get => address; set => address = value; }
        public string Email { get => email; set => email = value; }
        public string Phone { get => phone; set => phone = value; }
        public string FacultyId { get => facultyId; set => facultyId = value; }
        public string DepartmentId { get => departmentId; set => departmentId = value; }
        public bool State { get => state; set => state = value; }
        public ComeAndLeaveDate AdmissionDates { get => admissionDates; set => admissionDates = value; }
        public Salary Salaries { get => salaries; set => salaries = value; }
        // Constructors

        public Employee() {
            AdmissionDates = new ComeAndLeaveDate();
            Salaries = new Salary();
        }

        public Employee(string Id
        , string firstName
        , string lastName
        , int age
        , int gentle
        , string description
        , string address
        , string email
        , string phone
        , string facultyId
        , string departmentId
        , bool state
        , ComeAndLeaveDate admissionDates
        , Salary salaries
        )
        {
            this.Id = Id;
            this.firstName = firstName;
            this.lastName = lastName;
            this.age = age;
            this.gentle = gentle;
            this.description = description;
            this.address = address;
            this.email = email;
            this.phone = phone;
            this.facultyId = facultyId;
            this.departmentId = departmentId;
            this.state = state;
            this.admissionDates = admissionDates;
            this.salaries = salaries;
        }


        // Destructors
        ~Employee() { }
        // Methods
        // in, output
        public virtual void Input()
        {
            Console.Write("Employee Id: ");
            Id = Console.ReadLine();
            console.write("firstname: ");
            firstname = console.readline();
            console.write("lastname: ");
            lastname = console.readline();
            console.write("age: ");
            age = int32.parse(console.readline());
            console.write("gentle (0: male, 1: female): ");
            gentle = int32.parse(console.readline());
            console.write("description: ");
            description = console.readline();
            console.write("address: ");
            address = console.readline();
            console.write("email: ");
            email = console.readline();
            console.write("phone: ");
            phone = console.readline();
            console.write("facultyid: ");
            facultyid = console.readline();
            console.write("departmentid: ");
            departmentid = console.readline();
            console.write("state(1: free, other: 0): ");
            state = int32.parse(console.readline()) == 1 ? true : false;
            console.write("admissiondates: ");
            admissiondates.input();
            console.write("salaries: ");
            salaries.input();


        }
        public virtual void Output()
        {
            Console.WriteLine("Employee Id: {0}", Id);
            Console.WriteLine("FirstName: {0}", FirstName);
            Console.WriteLine("LastName: {0}", LastName);
            Console.WriteLine("Age: {0}", Age);
            Console.WriteLine("Gentle: {0}", Gentle);
            Console.WriteLine("Description: {0}", Description);
            Console.WriteLine("Address: {0}", Address);
            Console.WriteLine("Email: {0}", Email);
            Console.WriteLine("Phone: {0}", Phone);
            Console.WriteLine("FacultyId: {0}", FacultyId);
            Console.WriteLine("DepartmentId: {0}", DepartmentId);
            Console.WriteLine("State: {0}", State);
            Console.WriteLine("AdmissionDates:");
            AdmissionDates.Output();
            Console.WriteLine("Salaries: ");
            Salaries.Output();

        }

        // General method
        // Other method
        // Overriding
        public virtual void ExportBill()
        {

            Console.WriteLine($"The salary ( basis + bonus, allowance ): {CalcBill()}");
        }
        public virtual float CalcBill()
        => Salaries.Basic + Salaries.Bonus + Salaries.Allowance;








    }
}

File: src/hospitalManagement/Employee.java
package hospitalManagement;

import java.util.Scanner;
public abstract class Employee implements IBill 
    {
        //Field
        private String id;

        // Infor field
        private String firstName;

        private String lastName;

        private int age;

        private int gentle;

        private String description;

        private String address;

        private String email;

        private String phone;

        //  Management field
        private String facultyId;

        private String departmentId;

        private Boolean state;


        private ComeAndLeaveDate admissionDates;

        private Salary salaries;

        // Dynamic field
        static Scanner input = new Scanner(System.in);

        // Properties
        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public String getFirstName() {
            return firstName;
        }

        public void setFirstName(String firstName) {
            this.firstName = firstName;
        }

        public String getLastName() {
            return lastName;
        }

        public void setLastName(String lastName) {
            this.lastName = lastName;
        }

        public int getAge() {
            return age;
        }

        public void setAge(int age) {
            this.age = age;
        }

        public int getGentle() {
            return gentle;
        }

        public void setGentle(int gentle) {
            this.gentle = gentle;
        }

        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }

        public String getAddress() {
            return address;
        }

        public void setAddress(String address) {
            this.address = address;
        }

        public String getEmail() {
            return email;
        }

        public void setEmail(String email) {
            this.email = email;
        }

        public String getPhone() {
            return phone;
        }

        public void setPhone(String phone) {
            this.phone = phone;
        }

        public String getFacultyId() {
            return facultyId;
        }

        public void setFacultyId(String facultyId) {
            this.facultyId = facultyId;
        }

        public String getDepartmentId() {
            return departmentId;
        }

        public void setDepartmentId(String departmentId) {
            this.departmentId = departmentId;
        }

        public Boolean getState() {
            return state;
        }

        public void setState(Boolean state) {
            this.state = state;
        }

        public ComeAndLeaveDate getAdmissionDates() {
            return admissionDates;
        }

        public void setAdmissionDates(ComeAndLeaveDate admissionDates) {
            this.admissionDates = admissionDates;
        }

        public Salary getSalaries() {
            return salaries;
        }

        public void setSalaries(Salary salaries) {
            this.salaries = salaries;
        }
        
        // Constructors

        public Employee() { }

        public Employee (String Id
        , String firstName
        , String lastName
        , int age
        , int gentle
        , String description
        , String address
        , String email
        , String phone
        , String facultyId
        , String departmentId
        , Boolean state
        , ComeAndLeaveDate admissionDates
        , Salary salaries)
        {
            this.id = Id;
            this.firstName = firstName;
            this.lastName = lastName;
            this.age = age;
            this.gentle = gentle;
            this.description = description;
            this.address = address;
            this.email = email;
            this.phone = phone;
            this.facultyId = facultyId;
            this.departmentId = departmentId;
            this.state = state;
            this.admissionDates = admissionDates;
            this.salaries = salaries;
        }


        // Destructors
        // Methods
        // in, output
        public  void Input()
        {
            System.out.println("Employee Id: ");
            id = input.nextLine();
            System.out.println("FirstName: ");
            firstName = input.nextLine();
            System.out.println("LastName: ");
            lastName = input.nextLine();
            System.out.println("Age: ");
            age = input.nextInt();
            System.out.println("Gentle (0: male, 1: female): ");
            gentle = input.nextInt();
            System.out.println("Description: ");
            description = input.nextLine();
            System.out.println("Address: ");
            address = input.nextLine();
            System.out.println("Email: ");
            email = input.nextLine();
            System.out.println("Phone: ");
            phone = input.nextLine();
            System.out.println("FacultyId: ");
            facultyId = input.nextLine();
            System.out.println("DepartmentId: ");
            departmentId = input.nextLine();
            System.out.println("State(1: free, other: 0): ");
            state = input.nextInt() == 1 ? true : false;
            System.out.println("AdmissionDates: ");
            admissionDates.Input();
            System.out.println("Salaries: ");
            salaries.Input();

            
        }
        public void Output()
        {
            System.out.format("Employee Id: %s\n", id);
            System.out.format("FirstName: %s\n", firstName);
            System.out.format("LastName: %s\n}", lastName);
            System.out.format("Age: %d", age);
            System.out.format("Gentle: %d\n", gentle);
            System.out.format("Description: %s\n", description);
            System.out.format("Address: %s\n", address);
            System.out.format("Email: %s\n%s\n", email);
            System.out.format("Phone: %s\n", phone);
            System.out.format("FacultyId: %s\n", facultyId);
            System.out.format("DepartmentId: %s\n", departmentId);
            System.out.format("State: " + state +" \n");
            System.out.format("The patient list: ");
            System.out.format("AdmissionDates: %s\n", admissionDates);
            System.out.format("Salaries: " + salaries +" \n");

        }

        // General method
        // Other method
        // Overriding
        public void ExportBill()
        {

            System.out.format("The salary ( basis + bonus, allowance ): " + CalcBill() + " \n");
        }
        public float CalcBill(){
            return salaries.getBasic() + salaries.getBonus() + salaries.getAllowance();
        }

    }